module/zfs/zio_compress.c

```diff
@@ -148,5 +154,23 @@ zio_decompress_data(enum zio_compress c, abd_t *src, void *dst,
 	int ret = zio_decompress_data_buf(c, tmp, dst, s_len, d_len);
 	abd_return_buf(src, tmp, s_len);
 
+	/*
+	 * Decompression shouldn't fail, because we've already verifyied
+	 * the checksum.  However, for extra protection (e.g. against bitflips
+	 * in non-ECC RAM), we handle this error (and test it).
+	 */
+	ASSERT0(ret);
+	if (zio_decompress_fail_fraction != 0 &&
+	    spa_get_random(zio_decompress_fail_fraction) == 0)
+		ret = SET_ERROR(EINVAL);
```

Comment:
After taking a second look at this I think it's worth doing this decompress error injection with `zinject -e decompress`.  You can model it after the existing `zinject -e decrypt` support and the `zio_handle_decrypt_injection` handler used in `zio_decrypt`. 

This makes the code consistent and allows for an easy test case by adding a new `fault/decompress_fault.ksh` based on the existing `fault/decrypt_fault.ksh`.